### PR TITLE
Improve resiliency of entities when destroyed

### LINKED ISSUE #896: Crash when destroying subscription
**Description**
Presumably this occurs destroying any entity, but `rclnodejs` may crash if you destroy a subscription. Example below

- Library Version: `0.21.4`
- ROS Version: `foxy`
- Platform / OS: `20.04`
I'm using cyclonedds if that's relevant

**Steps To Reproduce**
```
const rclnodejs = require('rclnodejs');

function createSub1(node) {
  return node.createSubscription('std_msgs/msg/String', '/chatter1', msg => {
    console.log(`Chatter1 got: ${msg.data}`);
  });
}

function createSub1And2(node) {
  let sub1 = createSub1(node);

  let sub2 = node.createSubscription(
    'std_msgs/msg/String',
    '/chatter2',
    msg => {
      console.log(`Chatter2 got: ${msg.data}`);
      // process.nextTick(() => {
        // uncommenting the nextTick call (while still destroying/recreating) may change the behavior
        node.destroySubscription(sub1);
        sub1 = createSub1(node);
      // });
    },
  );
}

function spin(node) {
  setInterval(() => {
    node.spinOnce();
  }, 1);
}

async function main() {
  await rclnodejs.init();
  const node = new rclnodejs.Node('test_node');
  // node.spin();  // changing how spin is implemented changes the error
  spin(node);

  createSub1And2(node);

  const pub1 = node.createPublisher('std_msgs/msg/String', '/chatter1');
  const pub2 = node.createPublisher('std_msgs/msg/String', '/chatter2');

  let i = 0;
  setInterval(() => {
    pub1.publish({ data: 'Chatter1 ' + i++ });
    pub2.publish({ data: 'Chatter2 ' + i++ });
  }, 1);
}

main();
```

**Expected Behavior**
`rclnodejs` doesn't crash when destroying entities

**Actual Behavior**
`rclnodejs` may crash if entities are destroyed. My guess is that both subscriptions have data available and get marked as such. In the first callback the second subscription is destroyed. The executor then tries to execute the callback for a subscription that no longer exists.

There are a few comments in the code about small tweaks that change the outcome
1) using `node.spin` - this always ends in a segfault for me
```
// ... more message logs above
Chatter2 got: Chatter2 167
Chatter2 got: Chatter2 169
Chatter2 got: Chatter2 171
Segmentation fault (core dumped)
```
2) use `spin(node)` and destroy the subscription without `process.nextTick`- this almost immediately crashes with this error
```
Chatter1 got: Chatter1 0
Chatter2 got: Chatter2 1
Chatter2 got: Chatter2 3
rclnodejs/lib/node.js:191
          let success = rclnodejs.rclTake(subscription.handle, message);
                                  ^

Error: subscription pointer is invalid, at /tmp/binarydeb/ros-foxy-rcl-1.1.14/src/rcl/subscription.c:461
    at rclnodejs/lib/node.js:191:35
    at Node._runWithMessageType (rclnodejs/lib/node.js:1627:5)
    at rclnodejs/lib/node.js:188:12
    at Array.forEach (<anonymous>)
    at Node.execute (rclnodejs/lib/node.js:180:24)
    at Node.spinOnce (rclnodejs/lib/node.js:440:11)
 ```
3) use `spin(node)` and destroy the subscription in `process.nextTick` - this never crashes.

### PR DESCRIPTION
**Pull request by wayneparrott**
Issue #896 identifies an race condition where multiple subscriptions listen to a common topic and one subscription, from it's callback destroys another subscription (an entity). This revealed an issue where i/o activity can exists on an entity's _handle after the entity has been destroyed. The result is an invalid entity pointer or worse a SEGFAULT. To address this I've added destroy state semantics to the Entity class. 

full disclosure:
One quirky behavior I've attempted to resolve is that setting `Entity._handle = null`, results in the _handle being quickly GC'ed. This causes the SEGFAULT when the handle is still referenced by ShadowNode.Execute(). I believe after calling _handle.release(), the handle must not be GC'ed until after the current Node.execute() completes. I was not sure how to pull this off in C with Nan API so I implemented a hack by using a cache to hold `_handle`'s after their `release()`. The cache will flush when its size exceeds a const (100 atm). I'm definitely open to revising this hack with the suggestion of others.

entity.js
- added _destroy(), isDestroyed()
- added cache to hold obsolete _handles to avoid premature GC

entity.d.ts
- added isDestroyed()

node.js
- added guard statements to execute() to filter out destroyed entities

action/client.js
- added _destory() and isDestroyed()
- refactored destroy() flow

action/server.js
- added _destory() and isDestroyed()
- refactored destroy() flows

main.ts
- added tests for isDestroyed() to entity subclasses

test-destruction.js
- added entity.destory() test cases

Fix #896

### CODE REVIEW

File: lib/entity.js [hunk 1]
```diff
@@ -14,6 +14,19 @@
...
 'use strict';
 
+// Destroying an entity for which there is i/o activity can
+// results in a SEGFAULT. To mitigate this situation we hold a
+// reference to released handles to prevent them from being
+// GC'ed prematurely.
+const MAX_OBSOLETE_HANDLES_SIZE = 100;
+const OBSOLETE_HANDLES = new Set();
+function registerObsoleteHandle(handle) {
+  OBSOLETE_HANDLES.add(handle);
```
Comment (minggangw) [inline lib/entity.js:23]: I think we can check if the handle has been deleted when calling `execute`, and remove if the handle has been released.

https://github.com/RobotWebTools/rclnodejs/blob/0ea952f29e6251e229cf132790d61946ca4fca7f/lib/node.js#L150-L155

like
```js
execute(handles) {
  handles.forEach(handle => {
     if (OBSOLETE_HANDLES.indexOf(handle) !== -1) {
       // Remove from OBSOLETE_HANDLES.
      }
  });
...
```
I haven't tested, just in my mind:) thus we don't have to keep these handles for a long time.

**Comment by wayneparrott** [comment]
@minggangw my latest commit [9af468f](https://github.com/RobotWebTools/rclnodejs/commit/9af468f7bdc3ce4c7d0819c09d999e2162589b90) to address your suggestion may have overwritten [f3840e2](https://github.com/RobotWebTools/rclnodejs/commit/f3840e292e438bf21e3e84d187b18f64e2980caa)

File: lib/node.js [hunk 2]
```diff
@@ -371,7 +382,11 @@ class Node extends rclnodejs.ShadowNode {
         }
         GoalInfoArray.freeArray(message);
       }
-    });
+    }
+
+    // At this point it is safe to clear the cache of any
+    // destroyed entity references
+    Entity._gcHandles();
```
Comment (minggangw) [inline lib/node.js:389]: this looks pretty good :)

**Comment by minggangw** [comment]
> @minggangw my latest commit [9af468f](https://github.com/RobotWebTools/rclnodejs/commit/9af468f7bdc3ce4c7d0819c09d999e2162589b90) to address your suggestion may have overwritten [f3840e2](https://github.com/RobotWebTools/rclnodejs/commit/f3840e292e438bf21e3e84d187b18f64e2980caa)

Sorry for missing this comment, the change looks good to me, thanks!

**Review by minggangw — APPROVED** [review]
LGTM! Thanks for fixing the crashing issue

**Comment by minggangw** [comment]
Hi @wayneparrott I have approved this PR, please go ahead to merge it after resolving the conflict, thanks!

**Pull request merged**